overlord/servicestate/quota_handlers.go

```diff
@@ -92,46 +93,139 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
 
 	qc := qcs[0]
 
-	allGrps, err := AllQuotas(st)
+	updated, appsToRestartBySnap, err := quotaStateAlreadyUpdated(t)
 	if err != nil {
 		return err
 	}
 
-	var grp *quota.Group
-	switch qc.Action {
-	case "create":
-		grp, allGrps, err = quotaCreate(st, qc, allGrps)
-	case "remove":
-		grp, allGrps, err = quotaRemove(st, qc, allGrps)
-	case "update":
-		grp, allGrps, err = quotaUpdate(st, qc, allGrps)
-	default:
-		return fmt.Errorf("unknown action %q requested", qc.Action)
+	if !updated {
+		allGrps, err := AllQuotas(st)
+		if err != nil {
+			return err
+		}
+
+		var grp *quota.Group
+		switch qc.Action {
+		case "create":
+			grp, allGrps, err = quotaCreate(st, qc, allGrps)
+		case "remove":
+			grp, allGrps, err = quotaRemove(st, qc, allGrps)
+		case "update":
+			grp, allGrps, err = quotaUpdate(st, qc, allGrps)
+		default:
+			return fmt.Errorf("unknown action %q requested", qc.Action)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		// ensure service and slices on disk and their states are updated
+		opts := &ensureSnapServicesForGroupOptions{
+			allGrps: allGrps,
+		}
+		appsToRestartBySnap, err = ensureSnapServicesForGroup(st, t, grp, opts)
+		if err != nil {
+			return err
+		}
+
+		// after we have made all the persistent modifications
+		// to disk and state, remember that so to avoid
+		// failing redoing the non-idempotent parts of the
+		// task if snapd gets restarted before the end of the task.
```

Comment:
Could this be rephrased as follows (assuming that I got the right meaning):

"If snapd gets restarted before the end of the task, all the modifications done so far would have to be redone, and the non-idempotent parts of the task would fail. For this reason, once we have made all the persistent modifications to disk and state, we record the fact to avoid repeating them."

I'm sure that my English can be further improved :-)

Reply:
I reworded the first part a bit differently.